_posts/2025-02-13-rfprs.md

```diff
@@ -0,0 +1,372 @@
+level and your review mostly consists of style feedback, and perhaps some more nuanced feedback for
+the parts you are able to understand more easily.
+
+> This is clearly not the experience we want.
+
+As the creator of the PR, you are hoping to merge it. But past that, you also want your teammates to
+let you know of things that you may not have considered, or why this particular algorithm you've
+employed may have negative effects on the backend APIs it eventually calls. Style nits are OK but
+avoiding an incident by pointing out an unhandled edge case is much better.
+
+The reviewers only have so much time to devote to your PR. If that time is largely spent being
+mystified about the changeset, that is time that they are not spending giving you the quality
+feedback you should be hoping for. The time a reviewer spends on any particular PR is then
+multiplied by the number of people that end up reviewing it.
+
+> Time spent creating a better PR costs the business less to merge it.
+
+# RFPR Overview
+
+Reviewer First Pull Requests (RFPRs) are a prescriptive approach to building better PRs that are
+cheaper to review. At Mesosphere I worked on a team under [Kevin
+Klues](https://www.linkedin.com/in/klueska). After creating a PR which was hard to review he
+convinced me that his "Reviewer-First" methodology was worth adopting, and I've been promoting it
+ever since.
+
+> At the heart of RFPRs is the concept that PRs, like a book, should tell a story.
+
+When reading a book, you typically know what it's about before you start because you've read the
+intro on the back. You start at the first chapter, and continue reading until the end. Even if the
+book was particularly long or complex, the chapters were structured in a way to guide you through
+the story that the author wanted to tell.
+
+RFPRs operate under the same principle. A reviewer should be able to open the PR and read the
+description to get a good idea for what the PR is addressing and what to expect inside. RFPRs are
+structured so that they can be reviewed commit-by-commit, in order, leaving feedback as they go.
+
+> Large or complex PRs are made more approachable if the author guides the reviewers.
+
+# RFPR Guidelines
+
+Following are guidelines for creating RFPRs. Afterwards we'll discuss some tooling that will make
+following the guidelines much easier.
+
+## PR titles
+
+PR titles should describe what the PR accomplishes.
+
+Examples of good PR titles:
+
+- [CONTROL_42] Expose gRPC endpoints for all controlplane APIs.
+- kafka: Create new topic for user events and enable mirrormaker for all regions.
+- Modify CMS domain models to use composition instead of inheritance.
+
+PR titles which should be avoided:
+
+- [CONTROL_42] (just a JIRA ticket number)
+- fix bug
+- (nothing)
+
+Imagine that you are trying to find a particular PR on GitHub during an incident. The PR title
+should be descriptive enough so that it can be easily identified.
```

Comment:
Worth a mention or sharing your thoughts on [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)? I don't prescribe to it, but many people do so it may come up in comments.

Reply:
I don't actually have many thoughts on it. I've only used it at one company, and the adoption of it was pretty mixed, and I'm not sure I've ever gotten a lot of value out of it, so I don't think I could write anything meaningful about it.